javapoet/src/main/java/com/palantir/javapoet/TypeSpec.java

```diff
@@ -691,6 +741,52 @@ public Builder addSuperinterface(TypeMirror superinterface, boolean avoidNestedT
             return this;
         }
 
+        public Builder addRecordComponents(Iterable<ParameterSpec> parameterSpecs) {
+            checkArgument(parameterSpecs != null, "fieldSpecs == null");
+            for (ParameterSpec parameterSpec : parameterSpecs) {
+                addRecordComponent(parameterSpec);
+            }
+            return this;
+        }
+
+        public Builder addRecordComponent(ParameterSpec parameterSpec) {
+            recordComponents.add(parameterSpec);
+            return this;
+        }
+
+        public Builder varargs() {
+            return varargs(true);
+        }
+
+        public Builder varargs(boolean varargs) {
+            this.varargs = varargs;
+            return this;
+        }
+
+        /**
+         * Sets the compact constructor for this builder. Its parameters are solely used for javadoc
+         * generation.
+         */
+        public Builder compactConstructor(MethodSpec methodSpec) {
+            if (kind != Kind.RECORD) {
+                throw new UnsupportedOperationException(kind + " can't have compact constructors");
+            }
+
+            checkState(compactConstructor == null, "%s already has a compact constructor", name);
+
+            checkState(methodSpec.name().equals(MethodSpec.CONSTRUCTOR), "compact constructor is not a constructor");
+            EnumSet<Modifier> possibleModifiers = EnumSet.of(Modifier.PUBLIC, Modifier.PROTECTED, Modifier.PRIVATE);
+            checkState(
+                    possibleModifiers.containsAll(methodSpec.modifiers()),
+                    "compact constructor has non-access modifiers");
+            checkState(methodSpec.modifiers().size() <= 1, "compact constructor has too many access modifiers");
+            checkState(methodSpec.typeVariables().isEmpty(), "compact constructor has type variables");
+            checkState(methodSpec.exceptions().isEmpty(), "compact constructor has exceptions");
+            checkState(methodSpec.defaultValue() == null, "compact constructor has a default value");
```

Comment:
I don't want to expose `MethodSpec.CONSTRUCTOR` - that's an implementation detail of `MethodSpec`.

I'd prefer to avoid doing any validation of errors that can be caught by the compiler. There is a nearly limitless amount of validation we could do in these builders, but it comes at a cost:
- It makes the code significantly more complex and harder to understand
- We have to try to keep this validation up-to-date with the Java language specification

Instead I'd prefer we push the responsibility onto consumers to leverage the compiler for this validation. That way we know for certain that the validation matches the language specification.

```suggestion
```